### ISSUE TITLE
Things I tried as team lead: #1

### ISSUE THREAD

**Issue by HarshadRanganathan** [HarshadRanganathan/harshadranganathan.github.io#23]
https://martinfowler.com/articles/scaling-architecture-conversationally.html

**Comment by HarshadRanganathan** [comment]
https://medium.com/@crossbizz/let-service-teams-own-the-service-operations-instead-of-the-sre-4ff7bcbd53e0

**Comment by HarshadRanganathan** [comment]
https://engineering.mercari.com/en/blog/entry/20220121-introducing-platform-infra-team-at-mercari/

**Comment by HarshadRanganathan** [comment]
https://www.innoq.com/en/articles/2022/01/principles-of-technical-documentation/

**Comment by HarshadRanganathan** [comment]
https://martinfowler.com/articles/building-infrastructure-platform.html

**Comment by HarshadRanganathan** [comment]
Feedback form, Learning sessions, sharing credit, Rotation of scrum master, Rally updates, PR reminders

**Comment by HarshadRanganathan** [comment]
https://incident.io/blog/no-capes

**Comment by HarshadRanganathan** [comment]
https://docs.gruntwork.io/docs/guides/production-framework/

**Comment by HarshadRanganathan** [comment]
https://tech.deliveryhero.com/our-reliability-manifesto/

**Comment by HarshadRanganathan** [comment]
Video calls
Tech Lead Feedback
Learning Sessions
Sharing general updates, what's happening around us, future roadmap with the team

**Comment by HarshadRanganathan** [comment]
https://tech.deliveryhero.com/our-reliability-manifesto/

**Comment by HarshadRanganathan** [comment]
Achievements Journal,
Team Responsibilities

**Comment by HarshadRanganathan** [comment]
https://levelup.gitconnected.com/7-lessons-ive-learned-in-the-last-year-as-a-software-engineer-ff3a81e9dc13

**Comment by HarshadRanganathan** [comment]
https://www.adamblanchard.net/blog/personal-goal-setting

**Comment by HarshadRanganathan** [comment]
https://docs.moodle.org/311/en/Amazon_EC2_Cloud_Services_Installation

**Comment by HarshadRanganathan** [comment]
Work assignment, holiday plan
Assign work for each resource each day
Have some docs in backlog that could be picked up if blocked on any other issues

**Comment by HarshadRanganathan** [comment]
Support team with processes rather that as an individual e.g. docs

**Comment by HarshadRanganathan** [comment]
Discussion Points:

[1] Lack of motivation within the team
[2] Collaboration with external teams
[3] US delivery management without supervision/support
[4] Pre mortem goals
    Build architecture patterns for the cloud
    More automation around Infrastructure Provisioning
    Roll-out Well Architected framework for - AWS, Kubernetes
    Increased collaboration with other cloud teams
    Evolving Platform Team concept - Increased collaboration with Platform Tech Leads
    Building self-sufficient team

**Comment by HarshadRanganathan** [comment]
https://every.to/superorganizers/managing-your-manager

**Comment by HarshadRanganathan** [comment]
https://www.mckinsey.com/business-functions/people-and-organizational-performance/our-insights/if-were-all-so-busy-why-isnt-anything-getting-done

**Comment by HarshadRanganathan** [comment]
https://threadreaderapp.com/thread/1510285946285178887.html

**Comment by HarshadRanganathan** [comment]
https://swizec.com/blog/why-senior-engineers-get-nothing-done/

**Comment by HarshadRanganathan** [comment]
https://www.johnwhiles.com/posts/work.html

**Comment by HarshadRanganathan** [comment]
https://www.actitime.com/productivity/developer-productivity-tips

**Comment by HarshadRanganathan** [comment]
https://incident.io/blog/do-i-need-an-incident-debrief

**Comment by HarshadRanganathan** [comment]
https://blog.pragmaticengineer.com/performance-reviews-for-software-engineers/

**Comment by HarshadRanganathan** [comment]
https://dev.to/aws-heroes/architecture-devops-and-delivery-teams-need-to-think-differently-to-enable-serverless-12b1

**Comment by HarshadRanganathan** [comment]
https://blog.getambassador.io/is-platform-engineering-the-new-devops-or-sre-472ed97a1885

**Comment by HarshadRanganathan** [comment]
https://erik.wiffin.com/posts/how-to-get-the-most-out-of-your-11s

**Comment by HarshadRanganathan** [comment]
https://www.fastcompany.com/90719658/dont-quit-yet-this-is-how-to-make-your-current-job-fit-your-needs

**Comment by HarshadRanganathan** [comment]
https://www.freecodecamp.org/news/create-a-habit-system-and-stay-motivated-as-a-developer/

**Comment by HarshadRanganathan** [comment]
https://developers.google.com/tech-writing

**Comment by HarshadRanganathan** [comment]
https://mfdela.medium.com/in-defence-of-tech-debts-in-software-engineering-209685ae0e27

**Comment by HarshadRanganathan** [comment]
-  
- Teams to be trained on the technology they will be working on
- Teams to have at least one person who is an SME on topics e.g. Spark or have SME group for each topic/platform teams who will fill this need
- Dedicated team/guild/architect analysing and setting up standards/recommendations to be followed
- TPO’s working with business to have adequate timeline for iterative delivery to meet production standards
- Tech Leads pushing back on TPO to get adequate timeline for iterative delivery to meet production standards
- Not just trying to deliver fast to have nice metrics
- Not forcing teams with tight deadlines and moving to another work item without considering the fact that the previous work item needs to be maintained by someone on a recurring basis, whether it’s fully productionized etc.

 
Also, things to be considered for cloud migration –
 

- Someone has to manage the infrastructure in the cloud – moving to cloud doesn’t mean AWS will take care of everything
- Adequate funding/training/skillset needs to be given so that resources can stabilize the platform and avoid major P1 incidents down the line in terms of
- Having fine grained access controls in place
- Auditing
- Monitoring/Alerting
- Well architected design and many more
- Above shouldn’t require a justification – it needs to be quite obvious

**Comment by HarshadRanganathan** [comment]
https://www.linkedin.com/pulse/great-employees-dont-complain-walk-away-ian-daley

**Comment by HarshadRanganathan** [comment]
https://serverlessfirst.com/emails/quantifying-the-cost-of-introducing-extra-moving-parts-into-your-serverless-architecture/

**Comment by HarshadRanganathan** [comment]
Aspect | Standalone Application | Capability on Platform
-- | -- | --
Architecture and Security Reviews | Get proposed arch and infrastructure patterns reviewed and approved by Security Team | Adopt the existing security patterns and go for security review in case of any major arch changes
Infrastructure | Build new environments from the ground up | Leverage existing and provision additional resources as necessary
Implementation Strategy (CI/CD, Code, Test, etc.) | Set it up as you see fit for team. Can adopt patterns from the existing teams as well. | Embrace the practices from the platform engineering team and add additional ones if necessary
Code Starter Kits: | Build one news and/or adopt and customize existing ones from other teams | Adopt existing ones from platform teams and customize components if necessary
Community Support | Team will get up to speed on the UHG ecosystem, procedures and controls but can get additional support as needed | Help on offer from Platform team. Can't expect Platform team to support us in every step of the way though!
UHC.OPEN/Code reusability | Isolated applications have limited opportunities for inner source. Right mindset can still get things done. | Platform Engineering opens up synergies for collaboration and reuse.
Technology Upgrades | No dependency on other teams to pilot things. Free to execute disruptive things as long as business is not impacted | Collaborate with Platform team on new implementations. Need to take the entire platform ecosystem into consideration for impact!
Cost | It depends on what we can reuse? Should also take time invested by team in learning and building everything from the ground up | Reusability plays a big role in cost numbers
Speed to Market | It depends on what we can reuse? Should also take time invested by team in learning and building everything from the ground up | Reusability plays a big role in speed to market
Enterprise Direction | Isolated applications are still in use, depends on the use case and timeline. | Platform Engineering is the new normal!

**Comment by HarshadRanganathan** [comment]
https://martinfowler.com/articles/product-backlog-building-canvas.html

**Comment by HarshadRanganathan** [comment]
https://blog.pragmaticengineer.com/scaling-engineering-teams-via-writing-things-down-rfcs/

**Comment by HarshadRanganathan** [comment]
https://www.industrialempathy.com/posts/design-docs-at-google/

**Comment by HarshadRanganathan** [comment]
https://sharedphysics.com/everything-is-important/

**Comment by HarshadRanganathan** [comment]
https://sheeri.org/its-the-little-things/?

**Comment by HarshadRanganathan** [comment]
https://threadreaderapp.com/thread/1542061516912037890.html

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/178860613-e81dfe4a-7aea-4199-8ef1-d328ed991a7c.png)
![image](https://user-images.githubusercontent.com/32612787/178860617-59d14acc-ebf5-4280-8890-002af601eef1.png)

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/179117253-eee80ba1-cace-4166-940a-0e2188cfebe2.png)

**Comment by HarshadRanganathan** [comment]
https://www.infoq.com/podcasts/engineering-empathy

**Comment by HarshadRanganathan** [comment]
https://untools.co/

**Comment by HarshadRanganathan** [comment]
https://fs.blog/

**Comment by HarshadRanganathan** [comment]
https://medium.com/walmartglobaltech/building-a-platform-team-d915221d5654

**Comment by HarshadRanganathan** [comment]
https://rework.withgoogle.com/

**Comment by HarshadRanganathan** [comment]
https://slite.com/blog/micromanagement-is-not-a-bad-word

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/180989242-75b8afea-d36c-49ae-a192-5333809edf30.png)

**Comment by HarshadRanganathan** [comment]
https://www.infoq.com/articles/devops-governance-developer-velocity/

**Comment by HarshadRanganathan** [comment]
https://www.theengineeringmanager.com/qa/how-do-i-get-better-at-giving-feedback

**Comment by HarshadRanganathan** [comment]
https://docs.gruntwork.io/guides/production-framework/

**Comment by HarshadRanganathan** [comment]
https://blog.gruntwork.io/cloud-adoption-fails-65295aff30cc

**Comment by HarshadRanganathan** [comment]
https://blog.pragmaticengineer.com/oncall-compensation/

A: Oncall for software engineers is additional.

1. “Being oncall is your one and only job.”

2. “It’s not part of the job outside business hours.

3. “It’s not part of the job outside business hours, but we might still try to reach you during those times.”

4. “It’s part of the job for all software engineers and we operate in regions which regulate how it needs to be compensated with pay and time off.”

5. “It’s part of the job, but we recognize the disruption with pay and additional time off.”

6. “It’s voluntary for most people, and we encourage it with pay and time off.”

B: Oncall is part of the job:

7. “It’s part of the job for all software engineers and not paid additional.”

 
Being oncall can be quite disruptive in two major ways:
 
 
It disrupts your personal plans, outside of work.
It disrupts your sleep.
 
Compensation approaches 
 
Flat rate per week or per day of being oncall
Flat rate for standby, plus pay for hours worked outside core hours
Only pay for incidents worked on out-of-hours
 
 
Several engineers working at the company told me oncall operational load is high, teams are understaffed, oncall is not paid, and someone even used the term “oncall prison,” as quoted above. 
 
 
What is the reason for the high oncall lead? 
 
Growing too fast
Too many custom systems
Attrition for experienced people
No backfills
A barely acknowledged tech debt problem
Light at the end of the tunnel
 
Why are poor oncall practices painful? 
 
They can directly impact software engineer attrition and wellbeing. Simply put, poor oncall practices will lead to more engineers quitting, more people getting burnt out and fewer people recommending a company.
 
 
Takeaways
 
Oncall for software engineers is part of the job. Many companies operate like this, most notably Big Tech – save for Google – and many high-growth startups. The more an employer compensates software engineers, the more likely they expect oncall to be a given.
Oncall for software engineers is additional. Companies which care either about healthy oncall practices or want to minimize attrition for software engineers, make it clear oncall is additional and offer some sort of compensation. Compensation may be cash, or it could be time, or it could be lightening the load with dedicated SREs or DevOps people, or making the rotations voluntary.

**Comment by HarshadRanganathan** [comment]
https://www.docker.com/blog/building-stronger-happier-engineering-teams-with-team-topologies/

**Comment by HarshadRanganathan** [comment]
https://lucasfcosta.com/2022/08/07/how-to-improve-daily-standups.html

**Comment by HarshadRanganathan** [comment]
https://sifted.eu/articles/how-to-upskill-engineers/

**Comment by HarshadRanganathan** [comment]
https://architectelevator.com/cloud/serverless-design-patterns/

**Comment by HarshadRanganathan** [comment]
https://specbranch.com/posts/one-big-server/

**Comment by HarshadRanganathan** [comment]
https://holub.com/kpis-velocity-and-other-destructive-metrics/

**Comment by HarshadRanganathan** [comment]
https://longform.asmartbear.com/posts/extreme-questions/

**Comment by HarshadRanganathan** [comment]
https://www.nutshell.com/blog/accidental-complexity-software-design

**Comment by HarshadRanganathan** [comment]
https://medium.com/wise-engineering/platform-engineering-kpis-6a3215f0ee14

**Comment by HarshadRanganathan** [comment]
https://lucasfcosta.com/2022/08/07/how-to-improve-daily-standups.html

**Comment by HarshadRanganathan** [comment]
https://blog.ceejbot.com/posts/reduce-friction/

**Comment by HarshadRanganathan** [comment]
https://brooker.co.za/blog/2020/10/19/big-changes.html

**Comment by HarshadRanganathan** [comment]
https://talktotheduck.dev/external-debugging-tools-3-jmxterm

**Comment by HarshadRanganathan** [comment]
https://zwischenzugs.com/2022/08/08/who-should-write-the-terraform/

**Comment by HarshadRanganathan** [comment]
https://www.alexdebrie.com/posts/serverless-framework-vs-cdk/

**Comment by HarshadRanganathan** [comment]
https://www.priconceptions.com/notebook/remote-jobs-bad

**Comment by HarshadRanganathan** [comment]
<img width="863" alt="image" src="https://user-images.githubusercontent.com/32612787/188330169-51eb83bc-71f0-43cf-bf9e-af1aa28cbf73.png">

**Comment by HarshadRanganathan** [comment]
![Screenshot_20220914-232555_Outlook](https://user-images.githubusercontent.com/32612787/190273882-fdd53fd9-a15b-4797-9df0-de1ef76274a7.jpg)

**Comment by HarshadRanganathan** [comment]
![20220914_231923](https://user-images.githubusercontent.com/32612787/190273914-d0e86493-fe59-45c1-8086-058c886711a6.jpg)

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/190392606-acd2af4b-70a6-4348-a438-7d4d9feef0b1.png)

**Comment by HarshadRanganathan** [comment]
<p class="MsoNormal" style="margin: 0cm; font-size: 10pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><span style="font-size: 11pt;"><o:p> </o:p></span></p>

  | Ideal State | Current State | Problem Root Cause | Potential Solutions
-- | -- | -- | -- | --
Guild Meetings | Occurs as planned and gets cancelled occasionally due to overlaps with any other important meetings such as Town Hall | Sometimes guilds get cancelled in successive weeks without reason | Guild Owners are busy with their project delivery/other reasons/lack of agenda | Maybe Guilds should be run by a group of team members rather than one person/replacements
Agenda | Agenda is either well planned in advance or topics are gathered & discussed on the fly (lean meetings) | Sometimes no agenda/lack of audience/no active engagement from everyone | Guild Owners are busy with their project delivery/other reasons so no agenda sometimes Very less participation in the guilds due to lack of awareness/culture/tight delivery deadlines/time zone etc. | Below suggestion from Kevin Guild cohort should reach out to teams and build out an agenda based on their pain points/invite them to share their pain points/try different approaches and see what works
Purpose | Stays on track with the Guild’s mission/goal statements | Sometimes goes off-track discussing items which aren’t relevant to the purpose of that particular guild | Guild Owners are passionate to discuss about what interests them/lack of existing space to discuss things (busy calendar for everyone) so Guild is used as the place to discuss anything | Rename Guilds to make them more generic/send agenda in advance so only interested/relevant folks join/stay focused on the goals
Outcomes | Outcomes are assigned to Team members/Teams Outcomes are tracked Achieved outcomes are demoed | Some items no owner is identifiable | Lack of interest/lack of clarity as to who should own/deliver on it | Assign items to Platform teams backlog/interested teams/individuals – give them the needed support (moral + funding + time) as well
Participation | At least teams/members relevant to the guild participate | Siloed conversations  e.g. A team actively developing an API in our platform doesn’t participate in the guild  e.g. A team has already built a solution for a problem that the guild is actively discussing about | Lack of participation due to less awareness/culture/tight delivery deadlines/time zone etc. Lack of targeted communication | Project leads/TLs should encourage their teams to participate  Guild should ensure relevant scum teams are part of the conversation

<p class="MsoNormal" style="margin: 0cm; font-size: 10pt; font-family: Calibri, sans-serif; caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;"><span style="font-size: 11pt;"><o:p> </o:p></span></p>

**Comment by HarshadRanganathan** [comment]
Meetings:

- Everyone spoke and gave their opinions and as importantly were heard
- Everyone was interested and passionate about the subject
- Everyone had genuine positive intentions to make things better and solve problems
- Meeting notes (or a transcript) were captured
- Actionable items and clear takeaways were captured (with folks volunteering to take ownership of those takeaways)
- The meeting started on time and ended on time
- The phrase “that doesn’t make sense” was not used once (I am guilty of that one – have to work on my phrasing)
- At least two “dumb” questions were asked

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/199089317-5b101e7a-98d5-4f05-aa7c-510e15edc24a.png)

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/203998857-a1dff83a-eedb-4224-8d21-66ef92a5b433.png)

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/206041904-e27cc14d-e57b-4f1d-803f-b0fd1bd2aa0f.png)

**Comment by HarshadRanganathan** [comment]
![image](https://user-images.githubusercontent.com/32612787/206187972-ded2baa6-1b1c-477d-bc45-0b3f1aceed8d.png)
